### PR TITLE
Improve product selector data sync

### LINKED ISSUE #9200: Product selector - race condition affecting popular and recent products
Steps to reproduce:
1. Make sure the popular product is not among the first 25 products
2. Start order creation
3. Tap on Add Product
4. Notice that the popular section appears with the product
5. Tap on it and add it to the order -> notice the product is not added.
6. Tap on Add Product again
7. Notice the popular section is not visible.

The reason is that in step 3, the product is still in the database but as soon as we finish re-fetching the first page, the product is deleted from the local database -> when you select it we can’t add it to the order because it’s not in the local database anymore.

Result:
When I select a product from “Popular Products” it’s not added to the order.

### PR DESCRIPTION
**Pull request by samiuelson**
##### *Before merging this PR make sure the PR in FluxC is merged, then update FluxC version.

This PR changes products fetch operation in Product Selector to partial/mild sync style which does not wipe products table. This improves the performance of local search and fixes issues with the featured products not being added to an order.

Modification of data loading process:
1. Don't wipe the products table on the initial products load request.
2. Instead, display the initial page of existing results from the local db, and side re-fetch the first page in the background. When fetch completes, show the updated product list.
3. Paginate local db observable flow according to the current remote data fetch offset value. On each load-more request, fetch the next page of products and emit the next "page" from the local db. 

Closes: #9200

### Testing instructions

* Make sure you can add the featured product to order (from recent, and popular sections) in case the product is not present on the first page of products.
* Test if pagination works as expected

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/selector/ProductListHandler.kt [hunk 1]
```diff
@@ -18,7 +20,7 @@ class ProductListHandler @Inject constructor(private val repository: ProductSele
     }
 
     private val mutex = Mutex()
-    private var offset = 0
+    private var offset = MutableStateFlow(0)
```
Comment (samiuelson) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/selector/ProductListHandler.kt]: It's [thread-safe](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-mutable-state-flow/value.html) when we call it by the `value` prop.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/selector/ProductListHandler.kt [hunk 2]
```diff
@@ -28,21 +30,22 @@ class ProductListHandler @Inject constructor(private val repository: ProductSele
     private val productFilters = MutableStateFlow(mapOf<ProductFilterOption, String>())
 
     @OptIn(ExperimentalCoroutinesApi::class)
-    val productsFlow = combine(searchQuery, productFilters) { query, filters ->
+    val productsFlow: Flow<List<Product>> = combine(searchQuery, productFilters, offset) { query, filters, offset ->
```
Comment (malinajirka) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/selector/ProductListHandler.kt:34]: ❤️  Nice touch to explicitly define the type - makes it easier to read!

**Review by malinajirka — COMMENTED** [review]
Thanks @samiuelson, the code looks good to me, great job! I haven't noticed any issues. Initial tests worked fine, but I plan to test it more thoroughly after the final changes in FluxC are pushed.

**Comment by samiuelson** [comment]
@malinajirka, @AnirudhBhat It's ready for the final review.

**Review by malinajirka — APPROVED** [review]
LGTM, thanks @samiuelson !

Two things I'd consider for the future which are not strictly related to this PR:
1. Stop wiping the database even on the Product List screen - I'd personally wipe it only on pull-to-refresh.
2. Improve the SKU search so it searches in the local database and is consistent with the "name/description" search.

**Pull request merged**